modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -79,43 +105,115 @@ object GlobalL0Service {
+        logger.warn(e)(s"Failure pulling global snapshots from majority!") >>
+          Applicative[F].pure(
+            List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)]
+          )
+      }
+
+      private def pullSnapshots(
+        l0Peer: L0Peer,
+        nextOrdinal: SnapshotOrdinal,
+        lastOrdinal: SnapshotOrdinal
+      ): F[List[Hashed[GlobalIncrementalSnapshot]]] = {
+        val lastOrdinalCap = lastOrdinal.value.value
+          .min(singlePullLimit.map(nextOrdinal.value.value + _.value).getOrElse(lastOrdinal.value.value))
+
+        val ordinals = (nextOrdinal.value.value to lastOrdinalCap).toList
+          .map(o => SnapshotOrdinal(NonNegLong.unsafeFrom(o)))
+
+        (ordinals, List.empty[Hashed[GlobalIncrementalSnapshot]]).tailRecM {
+          case (ordinal :: nextOrdinals, snapshots) =>
+            l0GlobalSnapshotClient
+              .get(ordinal)(l0Peer)
+              .flatMap(_.toHashedWithSignatureCheck.flatMap(_.liftTo[F]))
+              .map(s => (nextOrdinals, snapshots :+ s).asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
+              .handleErrorWith { e =>
+                logger
+                  .warn(e)(s"Failure pulling snapshot with ordinal=$ordinal")
+                  .map(_ => snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
+              }
+
+          case (Nil, snapshots) =>
+            Applicative[F].pure(snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
+        }
+      }
+
+      private case class MajoritySnapshotData(ordinal: SnapshotOrdinal, hash: Hash)
+
+      // TODO should these majority functions move to MajorityPeerSelect.scala ?
+      private def getLatestMajoritySnapshotData(peerIds: NonEmptyList[PeerId]): F[MajoritySnapshotData] = {
+        val maybeData =
+          for {
+            peers <- OptionT[F, NonEmptyList[L0Peer]](
+              peerIds
+                .traverse(globalL0ClusterStorage.getPeer)
+                .map(_.toList.flatten.toNel)
+            )
+
+            majorityOrdinal <- OptionT[F, SnapshotOrdinal](
+              peers
+                .parTraverseN(10)(p => l0GlobalSnapshotClient.getLatestOrdinal.run(p))
+                .map(pickMajority[NonEmptyList, SnapshotOrdinal])
+            )
+
+            majorityHash <- OptionT[F, Hash](
+              peers.toList
+                .parTraverseN(10)(p => l0GlobalSnapshotClient.getHash(majorityOrdinal).run(p))
+                .map(_.flatten)
+                .map(pickMajority[List, Hash])
+            )
+          } yield MajoritySnapshotData(majorityOrdinal, majorityHash)
+
+        maybeData.getOrElseF(MonadThrow[F].raiseError(new Throwable("Unable to determine majority ordinal and hash")))
+      }
+
+      // TODO Random.shuffle initial list?
+      //  I did that first but it requires adding Random context bounds up the class hierarchy
```

Comment:
opening a discussion for this as well.
---
do you see any gain of randomizing the list? you need to query all of them anyway 🤔

Reply:
We don't query all of them -- we stop at the first peer that has the hash we're looking for

Reply:
ok, I see no issue to add `Random` boundary. anyway, I'm not sure it would be a gain. maybe we would distribute the load better not to overwhelm the peers on the beginning of the list 🤔

Reply:
Yeah I'm in favor here of randomness to help with the load problem.